### ISSUE TITLE
Truncated file names in the playlist

### ISSUE THREAD

**Issue by danulqua** [iina/iina#4641]
I have a TV Shows folder with different shows as folders which contains files as episodes. For example, folder "Halo" and files are named like "Halo (s01e01).mp4". If you open "Halo" folder in IINA, playlist will display file names without any issues like "Halo (s01e01).mp4" and so on. But if you open a single episode from that folder, IINA autocompletes the playlist with the rest of the files but with truncated names like "... 1).mp4", "... 2).mp4" and so on.

**System and IINA version:**

- macOS 14.0 (23A344)
- IINA 1.3.3 Build 138

**Expected behavior:**
If you open a single file in the folder, autocompleted playlist should have names "Halo (s01e01).mp4", "Halo (s01e02).mp4" and so on.

**Actual behavior:**
Autocompleted playlist consists of files with names "... 1).mp4", "... 2).mp4" and so on.

**Screenshots**
Open folder with IINA:
![CleanShot 2023-10-09 at 19 01 04@2x](https://github.com/iina/iina/assets/38916225/fd290030-9674-4108-a89f-649f369714cf)

Open single file whithin a folder:
![CleanShot 2023-10-09 at 18 59 44@2x](https://github.com/iina/iina/assets/38916225/5515add5-926b-4886-8ffe-ba99159eefa5)

**Steps to reproduce:**
1. Put some videos in a folder
2. Name these videos, for example, "Halo (s01e01)", "Halo (s01e02)" and so on
3. Open a single video from this folder
4. When IINA launches, open playlist
5. Notice that the file names are truncated

**How often does this happen?**
Always.

**Comment by danulqua** [comment]
Any feedback?

**Comment by svobs** [comment]
Recalling from memory... I too noticed this issue, and dove into the relevant code. There is a bunch of logic in the playlist item display which is clearly broken. But the few comments there were not helpful, and even the variable names  were not descriptive. Best guess was that it was intended to truncate the text when it got too small. It may have been designed before Apple's built-in line breaking functionality was available. In my fork I ended up just deleting a bunch of that logic and it was fixed. But I'm hesitant to submit a PR without knowing more about what was going on.

...OK, just took a peek at it in Interface Builder and this doesn't look good:
<img width="749" alt="SCR-20231020-ppzd" src="https://github.com/iina/iina/assets/2213815/406c9920-9b8d-4f10-9756-a02a3e0106c1">

I don't want to be the one to dive into this ATM though. Maybe @uiryuu can take a look when he's free?

**Comment by svobs** [comment]
OK, I went back and revisited this and was able to figure it out. It looks like this is not a bug, but actually a feature!

Perhaps to save horizontal space, or allow for easier comparison between files, IINA tries to figure out whether files in the playlist have identical text at the start of their titles, and replaces that text with `…`.

It looks like when a playlist is created, it scans all titles against each other and groups them by common prefix. It starts at the beginning of each title and specifically stops when it reaches a number. Interestingly, it looks for both [0-9] and for Chinese numbers (although I'm not yet clear on whether a number in the name is required to form a group). It's possible for it to form multiple groups, but there have to be at least 3 files with the same prefix to form a group. It then picks some number (?) of the biggest groups, and does the replacement for them.

You can actually click on the `…` and the name will be expanded to its full name. You can click on that segment again to collapse it.

An unusual feature, which looks like someone put a lot of work into. Might be pretty useful in some situations - I'll have to take some time and see if I can get used to it. Might be good if there was a switch to turn it off though.

**Comment by danulqua** [comment]
Indeed, I've never thought you can click the `...` to expand the title. Wasn't obvious at the first time, I'd like to have a switch for it, but if not, I can live with it.

**Comment by low-batt** [comment]
I've labeled this as a request for a setting to disable the feature that compacts the filename in the playlist.
